### PR TITLE
Add support for single node training with pt-xla

### LINKED ISSUE #1: Add support for single node training with pt-xla
Add support for single node training with pt-xla

See more detail at https://docs.google.com/document/d/18ZwHuRpCrWS5vqtIN7GSGtkwuLuT9HABXBXFuaK_fis/edit?resourcekey=0-qRBlZNdhusAZlbiAluAsFQ#heading=h.3mxn7yhd452y

### PR DESCRIPTION
**Pull request by thisisalbertliang**
Resolves #1

### CODE REVIEW

File: image_segmentation/pytorch/runtime/training.py [hunk 1]
```diff
@@ -41,7 +52,10 @@ def train(flags, model, train_loader, val_loader, loss_fn, score_fn, device, cal
         scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
                                                          milestones=flags.lr_decay_epochs,
                                                          gamma=flags.lr_decay_factor)
-    scaler = GradScaler()
+    if flags.torch_xla:
```
Comment (allenwang28) [inline image_segmentation/pytorch/runtime/training.py]: In this file in general, there are quite a few `if flags.torch_xla ... else` which starts to become difficult to read.

What do you think of refactoring these functions into a class (e.g. `class Trainer`) then having a `TorchXLATrainer` trainer?

File: image_segmentation/pytorch/runtime/training.py [hunk 2]
```diff
@@ -41,7 +52,10 @@ def train(flags, model, train_loader, val_loader, loss_fn, score_fn, device, cal
         scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
```
Comment (allenwang28) [inline image_segmentation/pytorch/runtime/training.py:81]: (GitHub won't let me comment in the right spot, but this comment is for your lines 47-48)
```
    torch.backends.cudnn.benchmark = flags.cudnn_benchmark
    torch.backends.cudnn.deterministic = flags.cudnn_deterministic
```

Should this change `if torch_xla`?

File: image_segmentation/pytorch/runtime/training.py [hunk 3]
```diff
@@ -91,6 +105,8 @@ def train(flags, model, train_loader, val_loader, loss_fn, score_fn, device, cal
                     scaler.update()
                 else:
                     optimizer.step()
+                    if flags.torch_xla:
+                        xm.mark_step()
```
Comment (miladm) [inline image_segmentation/pytorch/runtime/training.py]: @thisisalbertliang 
In light of the errors you've been observing:

I like your approach starts to focus on splitting the graph at iteration level. while this seem too fine grain, it's helpful to understand the code progress at finer granularity.

To better understand your code progress up to the training loop, consider adding a `mark_step()` before the `epoch` loop.

**Comment by thisisalbertliang** [comment]
The PR is updated to resolve [the previous pt-xla runtime error](https://paste.googleplex.com/4922837816049664) on TPU-VM.

With this change, UNet3D can now be successfully trained on TPU-VM (software version: `tpu-vm-pt-1.10`) using pt-xla with a single TPU core.

**Comment by miladm** [comment]
Thanks for the updates @thisisalbertliang. 
Looks like model can start to train on TPU VM fine. Looking forward to the TPU results.

**Pull request merged**